File: src/simulateur/SimulateurTEB.java
package simulateur;

import java.util.ArrayList;
import java.util.List;
import java.io.FileWriter;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Classe pour simuler la chaîne de transmission et générer la courbe de TEB en fonction du SNR.
 */
public class SimulateurTEB {

    private String typeModulation;
    private int nbSimulations; // Number of simulations per SNR
    private static final Boolean affichage = false;
    private AtomicInteger completedSimulations = new AtomicInteger(0); // Atomic integer for progress tracking
    
    /**
     * Constructeur de la classe SimulateurTEB.
     * Il prend en paramètre le type de modulation et le nombre de simulations.
     * @param typeModulation Type de modulation utilisé (NRZ, NRZT, RZ).
     * @param nbSimulations Nombre de simulations à effectuer pour chaque SNR.
     */
    public SimulateurTEB(String typeModulation, int nbSimulations) {
        this.typeModulation = typeModulation;
        this.nbSimulations = nbSimulations;
    }

    /**
     * Méthode pour générer la courbe de TEB en fonction du SNR, avec et sans codeur.
     * @param snrMin Le SNR minimum à tester.
     * @param snrMax Le SNR maximum à tester.
     * @param pasSNR Le pas entre les valeurs de SNR.
     * @param fichierCSV Le fichier dans lequel écrire les résultats.
     * @throws Exception Si une erreur survient pendant la simulation.
     */
    public void genererCourbeTEB(double snrMin, double snrMax, double pasSNR, String fichierCSV) throws Exception {
        List<Float> tebValuesWithoutCodeur = new ArrayList<>();
        List<Float> tebValuesWithCodeur = new ArrayList<>();
        List<Double> snrValues = new ArrayList<>();
        
        // Thread pool with a fixed number of threads based on available processors
        int availableThreads = Runtime.getRuntime().availableProcessors();
        ExecutorService executorService = Executors.newFixedThreadPool(availableThreads);
        
        // Calculate total simulations for progress tracking
        int totalSimulations = nbSimulations * (((int) ((snrMax - snrMin) / pasSNR)) + 1) * 2; // x2 for with and without codeur
        completedSimulations.set(0); // Reset the completed simulations count

        // Ouverture du fichier CSV pour écrire les résultats
        try (FileWriter writer = new FileWriter(fichierCSV)) {
            writer.write("Modulation,SNR(dB),TEB Without Codeur,TEB With Codeur\n"); // En-tête du fichier CSV

            // Boucle sur les valeurs de SNR
            List<Future<SimulationResult>> futures = new ArrayList<>();
            for (double snr = snrMin; snr <= snrMax; snr += pasSNR) {
                
                // Pour chaque SNR, on soumet les simulations au thread pool
                for (int simulation = 0; simulation < nbSimulations; simulation++) {
                    final int simIndex = simulation;
                    final double currentSnr = snr;
                    
                    // Simulations without the codeur
                    Future<SimulationResult> futureWithoutCodeur = executorService.submit(() -> {
                        Simulateur simulateur = new Simulateur(new String[] {
                            "-mess", "100", // Taille du message
                            "-form", typeModulation, // Type de modulation
                            "-seed", String.valueOf(simIndex + 1), // Germe différent pour chaque simulation
                            "-nbEch", "30", // Nombre d'échantillons par bit
                            "-ampl", "-1.0", "1.0", // Amplitude
                            "-snrpb", String.valueOf(currentSnr) // SNR par bit
                        });
                        
                        // Exécuter la simulation
                        simulateur.execute();
                        
                        // Calculer le TEB pour cette simulation
                        float teb = simulateur.calculTauxErreurBinaire();

                        // Update the progress bar
                        updateProgressBar(totalSimulations);

                        return new SimulationResult(currentSnr, teb, simIndex);
                    });
                    
                    futures.add(futureWithoutCodeur);

                    // Simulations with the codeur
                    Future<SimulationResult> futureWithCodeur = executorService.submit(() -> {
                        Simulateur simulateur = new Simulateur(new String[] {
                            "-mess", "1000", // Taille du message
                            "-form", typeModulation, // Type de modulation
                            "-seed", String.valueOf(simIndex + 1), // Germe différent pour chaque simulation
                            "-nbEch", "30", // Nombre d'échantillons par bit
                            "-ampl", "-1.0", "1.0", // Amplitude
                            "-snrpb", String.valueOf(currentSnr), // SNR par bit
                            "-codeur" // Activer le codeur
                        });
                        
                        // Exécuter la simulation
                        simulateur.execute();
                        
                        // Calculer le TEB pour cette simulation
                        float teb = simulateur.calculTauxErreurBinaire();

                        // Update the progress bar
                        updateProgressBar(totalSimulations);

                        return new SimulationResult(currentSnr, teb, simIndex);
                    });

                    futures.add(futureWithCodeur);
                }
            }
            
            // Récupérer les résultats des simulations
            for (int i = 0; i < futures.size(); i += 2) {
                SimulationResult resultWithoutCodeur = futures.get(i).get();   // Sans codeur
                SimulationResult resultWithCodeur = futures.get(i + 1).get(); // Avec codeur
                
                tebValuesWithoutCodeur.add(resultWithoutCodeur.teb);
                tebValuesWithCodeur.add(resultWithCodeur.teb);
                snrValues.add(resultWithoutCodeur.snr);
                
                if (affichage) {
                    System.out.println("Modulation: " + typeModulation + ", SNR: " + resultWithoutCodeur.snr + " dB, TEB Sans Codeur: " + resultWithoutCodeur.teb + ", TEB Avec Codeur: " + resultWithCodeur.teb);
                }
                
                // Écriture dans le fichier CSV
                writer.write(typeModulation + "," + resultWithoutCodeur.snr + "," + resultWithoutCodeur.teb + "," + resultWithCodeur.teb + "\n");
            }
            
            if (affichage) {
                System.out.println("Les résultats ont été enregistrés dans : " + fichierCSV);
            }
        } finally {
            // Shut down the executor service gracefully
            executorService.shutdown();
            executorService.awaitTermination(1, TimeUnit.HOURS);
        }
    }

    /**
     * Updates the global progress bar based on completed simulations.
     * @param totalSimulations The total number of simulations to be run.
     */
    private void updateProgressBar(int totalSimulations) {
        int completed = completedSimulations.incrementAndGet();
        int progress = (int) ((completed / (double) totalSimulations) * 100);
        System.out.print("\rProgress: " + progress + "% (" + completed + "/" + totalSimulations + " simulations)");
        if (completed == totalSimulations) {
            System.out.println("\nSimulation complete!");
        }
    }

    /**
     * Main method to run the simulation for different modulations and SNR values.
     * @param args Command line arguments.
     */
    public static void main(String[] args) {
        try {
            System.out.println("Simulation de la chaîne de transmission avec différents SNR pour les modulations NRZ, NRZT et RZ, avec et sans codeur");

            // Valeurs de SNR à tester
            double snrMin = -15.0;
            double snrMax = 10.0;
            double pasSNR = 1;

            int nbSimulations = 1; // Nombre de simulations pour chaque SNR

            // Créez les objets SimulateurTEB pour chaque modulation
            SimulateurTEB simTEBNRZ = new SimulateurTEB("NRZ", nbSimulations);
            SimulateurTEB simTEBNRZT = new SimulateurTEB("NRZT", nbSimulations);
            SimulateurTEB simTEBRZ = new SimulateurTEB("RZ", nbSimulations);

            // Générer les courbes TEB pour chaque modulation et enregistrer dans des fichiers CSV
            simTEBNRZ.genererCourbeTEB(snrMin, snrMax, pasSNR, "resultats/resultats_NRZ.csv");
            simTEBNRZT.genererCourbeTEB(snrMin, snrMax, pasSNR, "resultats/resultats_NRZT.csv");
            simTEBRZ.genererCourbeTEB(snrMin, snrMax, pasSNR, "resultats/resultats_RZ.csv");

            System.out.println("Fin de la simulation de la chaîne de transmission");

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    // Classe pour encapsuler le résultat d'une simulation
    private static class SimulationResult {
        double snr;
        float teb;
        int simulationIndex;

        public SimulationResult(double snr, float teb, int simulationIndex) {
            this.snr = snr;
            this.teb = teb;
            this.simulationIndex = simulationIndex;
        }
    }
}

File: simulateur_teb.py
import subprocess
import re
import csv
from concurrent.futures import ThreadPoolExecutor, as_completed
from threading import Lock

class SimulateurTEB:
    def __init__(self, type_modulation, nb_simulations):
        """
        Constructor for SimulateurTEB class.

        :param type_modulation: Modulation type used (NRZ, NRZT, RZ)
        :param nb_simulations: Number of simulations to perform for each SNR value
        """
        self.type_modulation = type_modulation
        self.nb_simulations = nb_simulations
        self.lock = Lock()  # For thread-safe progress update
        self.completed_simulations = 0

    def run_simulation(self, mess, snr, use_codeur, nbEch=30):
        """
        Runs a single simulation using the 'simulateur' executable and returns the TEB value.

        :param mess: Message length (int)
        :param snr: Signal-to-noise ratio per bit (float)
        :param use_codeur: Whether to use the codeur (True/False)
        :param nbEch: Number of samples per bit (int)
        :return: A tuple with the current SNR and the TEB value
        """
        command = ['./simulateur', '-mess', str(mess), '-form', self.type_modulation, '-snrpb', str(snr), '-nbEch', str(nbEch)]
        if use_codeur:
            command.append('-codeur')

        try:
            result = subprocess.run(command, capture_output=True, text=True, check=True)
            output = result.stdout

            # Extract the TEB value using regex (handle scientific notation like 1.0E-4)
            teb_match = re.search(r'TEB\s*:\s*([+-]?\d*\.?\d+[eE]?[+-]?\d*)', output)
            if teb_match:
                teb_value = float(teb_match.group(1))
                return teb_value
            else:
                print(f"TEB value not found in the output for SNR = {snr}")
                return None

        except subprocess.CalledProcessError as e:
            print(f"Error during execution for SNR = {snr}: {e}")
            return None

    def generer_courbe_TEB(self, snr_min, snr_max, pas_snr, fichier_csv, mess):
        """
        Generates a curve of TEB as a function of SNR, with and without the codeur.

        :param snr_min: Minimum SNR value (float)
        :param snr_max: Maximum SNR value (float)
        :param pas_snr: Step size between SNR values (float)
        :param fichier_csv: Path to CSV file for saving the results
        :param mess: Message length for the simulation (int)
        """
        snr_values = [snr_min + i * pas_snr for i in range(int((snr_max - snr_min) / pas_snr) + 1)]
        total_simulations = len(snr_values) * self.nb_simulations * 2  # *2 for with and without codeur
        futures = []

        # Create thread pool to run multiple simulations concurrently
        with ThreadPoolExecutor(max_workers=4) as executor, open(fichier_csv, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(['Modulation', 'SNR(dB)', 'TEB Without Codeur', 'TEB With Codeur'])  # CSV Header

            for snr in snr_values:
                for _ in range(self.nb_simulations):
                    # Submit both simulations (with and without codeur)
                    future_without = executor.submit(self.run_simulation, mess, snr, False)  # Without codeur
                    future_with = executor.submit(self.run_simulation, mess, snr, True)   # With codeur

                    futures.append((snr, future_without, future_with))

            # Collect the results as they complete
            for future_snr, future_without_codeur, future_with_codeur in futures:
                teb_without_codeur = future_without_codeur.result()
                teb_with_codeur = future_with_codeur.result()

                if teb_without_codeur is not None and teb_with_codeur is not None:
                    # Write both results for the same SNR to the CSV
                    writer.writerow([self.type_modulation, future_snr, teb_without_codeur, teb_with_codeur])

                # Update progress for each completed pair
                with self.lock:
                    self.completed_simulations += 2  # Two simulations done per SNR value
                    self.update_progress(self.completed_simulations, total_simulations)

        print(f"Results saved to {fichier_csv}")

    def update_progress(self, completed, total):
        """
        Updates the progress bar.

        :param completed: Completed simulations count
        :param total: Total number of simulations
        """
        progress = int((completed / total) * 100)
        print(f"\rProgress: {progress}% ({completed}/{total})", end='')


if __name__ == "__main__":
    # Example usage of the SimulateurTEB class
    snr_min = -10.0
    snr_max = 20.0
    pas_snr = 0.1
    mess = 100000
    nb_simulations = 1

    # Create instances for each modulation type
    simTEBNRZ = SimulateurTEB("NRZ", nb_simulations)
    simTEBNRZT = SimulateurTEB("NRZT", nb_simulations)
    simTEBRZ = SimulateurTEB("RZ", nb_simulations)

    # Generate TEB curves and save to CSV
    simTEBNRZ.generer_courbe_TEB(snr_min, snr_max, pas_snr, 'resultats/resultats_NRZ.csv', mess)
    simTEBNRZT.generer_courbe_TEB(snr_min, snr_max, pas_snr, 'resultats/resultats_NRZT.csv', mess)
    simTEBRZ.generer_courbe_TEB(snr_min, snr_max, pas_snr, 'resultats/resultats_RZ.csv', mess)

    print("Simulation complete")
